src/fluree/raft.clj

```diff
@@ -52,20 +58,49 @@
 
 (defn close
   "Closes a raft process."
-  [{:keys [config] :as raft}]
-  (log/info "Shutting down raft")
+  [{:keys [config this-server] :as raft}]
+  (log/info (format "Server %s shutting down raft" this-server))
   (let [{:keys [snapshot-write]} config]
     (when (fn? snapshot-write)
       (let [commit (-> raft raft-state-async async/<!! :commit)]
         (snapshot-write commit
-                        #(log/info "Wrote final snapshot at for index" commit)))))
+                        #(log/info
+                           (format "Server %s wrote final snapshot for index %s"
+                                   this-server commit))))))
   (let [close-fn (:close-fn config)]
     (async/close! (events/event-chan raft))
     (if (fn? close-fn)
       (close-fn)
       :closed)))
 
 
+(defn add-server
+  "Adds an already created server to the raft network. You can create the
+  server by calling `start` with the appropriate config, then pass its id to
+  this fn in the `new-server-id` arg. Due to the way the raft protocol works,
+  you can only add one server at a time."
+  [raft new-server-id]
+  (raft-specs/valid-or-throw ::raft-specs/raft raft "invalid raft state")
```

Comment:
I think we might be better off with an assertion here as I don't think user input would change this. Assuming this is primarily for development validation, an assertion would catch errors during development but not unnecessarily slow down production.

Reply:
Yeah, I went back and forth on this too. I wouldn't want to call `valid-or-throw` anywhere that is invoked often or in a very performance-critical section of code. I figured `add-server` was sufficiently seldom invoked and not performance sensitive enough for this to make a noticeable difference. But what do you think?

If we do start adding these checks anywhere that would negatively impact performance in production, then I agree that we should use assertions we turn off in production.

Reply:
Agree probably not an issue here. Concern would be less about frequency of calls and more on Raft timeouts if this operation is blocking a broadcast message (synchronous inside the heartbeat). With a sufficiently low heartbeat timeout, a millisecond can make a difference.

Reply:
Makes sense. I'll investigate a bit before merging.